Matching on `tea.Msg` results in incorrect Go output

Lisette version: 0.1.15

The following Lisette file compiles but results in unexpected Go output for the `Update` method:
```rs
import "go:charm.land/bubbletea/v2"
import "go:fmt"

struct Model { width: int, height: int }

impl Model {
  fn Init(self: Model) -> Option<tea.Cmd> {
    None
  }

  fn Update(self: Model, msg: tea.Msg) -> (tea.Model, Option<tea.Cmd>) {
    match msg {
      tea.WindowSizeMsg { Width: width, Height: height } => (
        Model { width, height, ..self },
        None,
      ),
      tea.KeyPressMsg { Text: text, .. } => {
        fmt.Println(text)
        (self, None)
      },
      _ => (self, None),
    }
  }

  fn View(self: Model) -> tea.View {
    tea.NewView(fmt.Sprintf("%dx%d", self.width, self.height))
  }
}

fn main() {
  if let Err(err) = tea.NewProgram(Model { width: 0, height: 0 }).Run() {
    panic(err)
  }
}
```
This is the Go output that is generated for the `Update` method above:
```go
func (m Model) Update(msg tea.Msg) lisette.Tuple2[tea.Model, lisette.Option[tea.Cmd]] {
	width := msg.Width
	height := msg.Height
	copy_1 := m
	copy_1.width = width
	copy_1.height = height
	return lisette.MakeTuple2[tea.Model, lisette.Option[tea.Cmd]](_lisAdapter_Model_Model_0{inner: copy_1}, lisette.MakeOptionNone[tea.Cmd]())
}
```
It has the following two errors:
> ./main.go:24:15: msg.Width undefined (type tea.Msg has no field or method Width)
./main.go:25:16: msg.Height undefined (type tea.Msg has no field or method Height)

It seems like this generated code is unwrapping the contents of that first match branch and discarding the rest of the function, perhaps because it thinks this first branch covers all possibilities (even though that issue was fixed for the LSP/compiler).

--- a/crates/emit/src/go/definitions/interface_adapter.rs
+++ b/crates/emit/src/go/definitions/interface_adapter.rs
@@ -71,19 +71,11 @@ impl Emitter<'_> {
                 return Some(under);
             }
         }
-        let Type::Constructor { id, .. } = &resolved else {
-            return None;
-        };
-        match self.ctx.definitions.get(id.as_str()) {
-            Some(Definition::TypeAlias { ty: alias_ty, .. }) => {
-                let aliased = alias_ty.resolve();
-                if matches!(aliased, Type::Function { .. }) {
-                    Some(aliased)
-                } else {
-                    None
-                }
-            }
-            _ => None,
+        let peeled = self.peel_alias(&resolved);
+        if matches!(peeled, Type::Function { .. }) {
+            Some(peeled)
+        } else {
+            None
         }
     }
 
@@ -98,7 +90,7 @@ impl Emitter<'_> {
                 }
             }
             for parent_ty in &current.parents {
-                let parent = parent_ty.resolve();
+                let parent = self.peel_alias(parent_ty);
                 let Type::Constructor { id, .. } = &parent else {
                     continue;
                 };
@@ -115,7 +107,7 @@ impl Emitter<'_> {
     }
 
     pub(crate) fn needs_adapter(&self, source_ty: &Type, target_ty: &Type) -> Option<AdapterPlan> {
-        let target = target_ty.resolve();
+        let target = self.peel_alias(target_ty);
         let Type::Constructor { id: target_id, .. } = &target else {
             return None;
         };

--- a/crates/emit/src/go/expressions/dot_classify.rs
+++ b/crates/emit/src/go/expressions/dot_classify.rs
@@ -414,13 +414,10 @@ impl Emitter<'_> {
             return None;
         }
 
-        let (qualified_type, _type_name) = if let Definition::TypeAlias {
-            ty: Type::Constructor { id, .. },
-            ..
-        } = definition
-        {
-            let resolved_name = id.rsplit('.').next().unwrap_or(id);
-            (id.to_string(), resolved_name.to_string())
+        let (qualified_type, _type_name) = if matches!(definition, Definition::TypeAlias { .. }) {
+            let id = self.peel_alias_id(&qualified_type);
+            let resolved_name = id.rsplit('.').next().unwrap_or(&id).to_string();
+            (id, resolved_name)
         } else {
             (qualified_type, type_name.to_string())
         };

--- a/crates/emit/src/go/expressions/values.rs
+++ b/crates/emit/src/go/expressions/values.rs
@@ -360,7 +360,7 @@ impl Emitter<'_> {
     ) -> String {
         let inner = self.emit_operand(output, expression);
 
-        if let Type::Constructor { id, .. } = &ty.resolve()
+        if let Type::Constructor { id, .. } = &self.peel_alias(ty)
             && matches!(
                 self.ctx.definitions.get(id.as_str()),
                 Some(Definition::Interface { .. })

--- a/crates/emit/src/go/names/generics.rs
+++ b/crates/emit/src/go/names/generics.rs
@@ -111,7 +111,10 @@ impl Emitter<'_> {
         if !visited.insert(id.to_string()) {
             return HashSet::default();
         }
-        let Some(Definition::Interface { definition, .. }) = self.ctx.definitions.get(id) else {
+        let peeled = self.peel_alias_id(id);
+        let Some(Definition::Interface { definition, .. }) =
+            self.ctx.definitions.get(peeled.as_str())
+        else {
             return HashSet::default();
         };
         let names: Vec<&str> = definition

--- a/crates/emit/src/go/names/resolution.rs
+++ b/crates/emit/src/go/names/resolution.rs
@@ -1,7 +1,5 @@
 use crate::Emitter;
 use crate::go::names::go_name;
-use syntax::program::Definition;
-use syntax::types::Type;
 
 impl Emitter<'_> {
     pub(crate) fn resolve_go_name(&mut self, name: &str) -> String {
@@ -40,16 +38,15 @@ impl Emitter<'_> {
 
     pub(crate) fn resolve_alias_type_name(&self, type_part: &str) -> Option<String> {
         let qualified = format!("{}.{}", self.current_module, type_part);
-        if let Some(Definition::TypeAlias { ty, .. }) = self.ctx.definitions.get(qualified.as_str())
-            && let Type::Constructor { id, .. } = &ty.resolve()
-        {
-            let type_module = id.split('.').next().unwrap_or("");
-            if type_module == self.current_module {
-                return Some(id.split('.').next_back()?.to_string());
-            }
-            return Some(id.to_string());
+        let id = self.peel_alias_id(&qualified);
+        if id == qualified {
+            return None;
         }
-        None
+        let type_module = id.split('.').next().unwrap_or("");
+        if type_module == self.current_module {
+            return Some(id.split('.').next_back()?.to_string());
+        }
+        Some(id)
     }
 
     pub(crate) fn capitalize_static_method_if_public(&self, name: &str) -> String {

--- a/crates/emit/src/go/queries/definitions.rs
+++ b/crates/emit/src/go/queries/definitions.rs
@@ -30,9 +30,9 @@ impl Emitter<'_> {
     }
 
     pub(crate) fn field_is_public(&self, struct_ty: &Type, field_name: &str) -> bool {
-        let resolved = struct_ty.resolve();
+        let resolved = self.peel_alias(&struct_ty.resolve().strip_refs());
 
-        let Type::Constructor { id, .. } = resolved.strip_refs() else {
+        let Type::Constructor { id, .. } = &resolved else {
             return false;
         };
 
@@ -154,8 +154,48 @@ impl Emitter<'_> {
         None
     }
 
+    pub(crate) fn peel_alias(&self, ty: &Type) -> Type {
+        let mut current = ty.resolve();
+        let mut seen: Vec<String> = Vec::new();
+        loop {
+            let Type::Constructor { id, .. } = &current else {
+                return current;
+            };
+            if seen.iter().any(|s| s == id.as_str()) {
+                return current;
+            }
+            let Some(Definition::TypeAlias { ty: alias_ty, .. }) =
+                self.ctx.definitions.get(id.as_str())
+            else {
+                return current;
+            };
+            seen.push(id.to_string());
+            current = alias_ty.resolve();
+        }
+    }
+
+    pub(crate) fn peel_alias_id(&self, id: &str) -> String {
+        let mut current = id.to_string();
+        let mut seen: Vec<String> = Vec::new();
+        loop {
+            if seen.iter().any(|s| s == &current) {
+                return current;
+            }
+            let Some(Definition::TypeAlias { ty: alias_ty, .. }) =
+                self.ctx.definitions.get(current.as_str())
+            else {
+                return current;
+            };
+            let Type::Constructor { id: next, .. } = alias_ty.resolve() else {
+                return current;
+            };
+            seen.push(current);
+            current = next.to_string();
+        }
+    }
+
     pub(crate) fn as_enum(&self, ty: &Type) -> Option<String> {
-        let Type::Constructor { id, .. } = ty.resolve() else {
+        let Type::Constructor { id, .. } = self.peel_alias(ty) else {
             return None;
         };
 
@@ -170,7 +210,7 @@ impl Emitter<'_> {
     }
 
     pub(crate) fn as_interface(&self, ty: &Type) -> Option<String> {
-        let Type::Constructor { id, .. } = ty.resolve() else {
+        let Type::Constructor { id, .. } = self.peel_alias(ty) else {
             return None;
         };
 

--- a/tests/spec/emit/go_interface_adapter.rs
+++ b/tests/spec/emit/go_interface_adapter.rs
@@ -580,3 +580,75 @@ fn main() {
 "#;
     assert_emit_snapshot_with_go_typedefs!(input, &[]);
 }
+
+#[test]
+fn cast_to_aliased_go_interface_applies_adapter() {
+    let input = r#"
+import "go:example.com/svc"
+
+struct Loader {}
+
+impl Loader {
+  fn Load(self, key: string) -> Result<int, error> {
+    Ok(1)
+  }
+}
+
+fn run(s: svc.Alias) -> int {
+  match s.Load("k") {
+    Ok(n) => n,
+    Err(_) => 0,
+  }
+}
+
+fn main() {
+  let l = Loader {}
+  let _ = run(l as svc.Alias)
+}
+"#;
+    let typedef = r#"
+pub interface Service {
+  fn Load(key: string) -> Result<int, error>
+}
+pub type Alias = Service
+"#;
+    assert_emit_snapshot_with_go_typedefs!(input, &[("go:example.com/svc", typedef)]);
+}
+
+#[test]
+fn adapter_collects_methods_from_aliased_parent_interface() {
+    let input = r#"
+import "go:example.com/shapes"
+
+struct Square {}
+
+impl Square {
+  fn Area(self) -> Result<int, error> {
+    Ok(4)
+  }
+  fn Name(self) -> string {
+    "sq"
+  }
+}
+
+fn describe(s: shapes.Shape) -> string {
+  s.Name()
+}
+
+fn main() {
+  let s = Square {}
+  let _ = describe(s as shapes.Shape)
+}
+"#;
+    let typedef = r#"
+pub interface Sized {
+  fn Area() -> Result<int, error>
+}
+pub type SizedAlias = Sized
+pub interface Shape {
+  impl SizedAlias
+  fn Name() -> string
+}
+"#;
+    assert_emit_snapshot_with_go_typedefs!(input, &[("go:example.com/shapes", typedef)]);
+}

--- a/tests/spec/emit/interop_matrix.rs
+++ b/tests/spec/emit/interop_matrix.rs
@@ -1118,3 +1118,24 @@ pub type Cmd = fn() -> Event
 "#;
     assert_emit_snapshot_with_go_typedefs!(input, &[("go:example.com/scheduler", typedef)]);
 }
+
+#[test]
+fn interop_option_of_aliased_interface_uses_is_nil_interface() {
+    let input = r#"
+import "go:example.com/evts"
+import "go:fmt"
+
+fn main() {
+  match evts.Peek() {
+    Some(e) => fmt.Println(e),
+    None => fmt.Println("none"),
+  }
+}
+"#;
+    let typedef = r#"
+pub interface Event {}
+pub type Msg = Event
+pub fn Peek() -> Option<Msg>
+"#;
+    assert_emit_snapshot_with_go_typedefs!(input, &[("go:example.com/evts", typedef)]);
+}

--- a/tests/spec/emit/patterns.rs
+++ b/tests/spec/emit/patterns.rs
@@ -628,3 +628,66 @@ pub struct Resize { pub width: int, pub height: int }
 "#;
     assert_emit_snapshot_with_go_typedefs!(input, &[("go:example.com/events", typedef)]);
 }
+
+#[test]
+fn match_on_aliased_go_interface_emits_type_switch() {
+    let input = r#"
+import "go:example.com/events"
+
+fn handle(m: events.Msg) -> int {
+  match m {
+    events.Click { x, y } => x + y,
+    events.KeyPress { key } => key.length(),
+    _ => 0,
+  }
+}
+"#;
+    let typedef = r#"
+pub interface Event {}
+pub type Msg = Event
+pub struct Click { pub x: int, pub y: int }
+pub struct KeyPress { pub key: string }
+"#;
+    assert_emit_snapshot_with_go_typedefs!(input, &[("go:example.com/events", typedef)]);
+}
+
+#[test]
+fn match_on_aliased_lisette_enum_emits_enum_tag_switch() {
+    let input = r#"
+enum Color { Red, Green, Blue }
+
+type Palette = Color
+
+fn describe(p: Palette) -> string {
+  match p {
+    Color.Red => "r",
+    Color.Green => "g",
+    Color.Blue => "b",
+  }
+}
+"#;
+    assert_emit_snapshot!(input);
+}
+
+#[test]
+fn match_on_chained_alias_of_go_interface_emits_type_switch() {
+    let input = r#"
+import "go:example.com/events"
+
+fn handle(m: events.Outer) -> int {
+  match m {
+    events.Click { x, y } => x + y,
+    events.KeyPress { key } => key.length(),
+    _ => 0,
+  }
+}
+"#;
+    let typedef = r#"
+pub interface Event {}
+pub type Inner = Event
+pub type Outer = Inner
+pub struct Click { pub x: int, pub y: int }
+pub struct KeyPress { pub key: string }
+"#;
+    assert_emit_snapshot_with_go_typedefs!(input, &[("go:example.com/events", typedef)]);
+}

